File: engine/src/main/java/com/bartlomiejpluta/base/core/gl/render/Renderer.java
package com.bartlomiejpluta.base.core.gl.render;

import com.bartlomiejpluta.base.core.gc.Cleanable;
import com.bartlomiejpluta.base.core.ui.Window;
import com.bartlomiejpluta.base.core.world.camera.Camera;

public interface Renderer extends Cleanable {
   void init();
   void render(Window window, Camera camera, Renderable renderable);
}

File: editor/src/main/kotlin/com/bartlomiejpluta/base/editor/view/render/renderer/Renderer.kt
package com.bartlomiejpluta.base.editor.view.render.renderer

import com.bartlomiejpluta.base.editor.render.model.Renderable
import javafx.animation.AnimationTimer
import javafx.scene.canvas.GraphicsContext

class Renderer(
    private val gc: GraphicsContext,
    private val renderable: Renderable
) : AnimationTimer() {
    private var previous = System.nanoTime()

    override fun handle(now: Long) {
        val dt = (now - previous) / 1000000000.0
        previous = now
        gc.isImageSmoothing = false

        render()
    }

    private fun render() {
        gc.clearRect(0.0, 0.0, gc.canvas.width, gc.canvas.height);

        renderable.render(gc)
    }
}